sklearn/utils/_metadata_requests.py

```diff
@@ -0,0 +1,1253 @@
+        of the method.
+
+    Notes
+    -----
+    This class is a descriptor [1]_ and uses PEP-362 to set the signature of
+    the returned function [2]_.
+
+    References
+    ----------
+    .. [1] https://docs.python.org/3/howto/descriptor.html
+
+    .. [2] https://www.python.org/dev/peps/pep-0362/
+    """
+
+    def __init__(self, name, keys, validate_keys=True):
+        self.name = name
+        self.keys = keys
+        self.validate_keys = validate_keys
+
+    def __get__(self, instance, owner):
+        # we would want to have a method which accepts only the expected args
+        def func(**kw):
+            """Updates the request for provided parameters
+
+            This docstring is overwritten below.
+            See REQUESTER_DOC for expected functionality
+            """
+            if self.validate_keys and (set(kw) - set(self.keys)):
+                raise TypeError(
+                    f"Unexpected args: {set(kw) - set(self.keys)}. Accepted arguments"
+                    f" are: {set(self.keys)}"
+                )
+
+            requests = instance._get_metadata_request()
+            method_metadata_request = getattr(requests, self.name)
+
+            for prop, alias in kw.items():
+                if alias is not UNCHANGED:
+                    method_metadata_request.add_request(param=prop, alias=alias)
+                    requests._is_default_request = False
+            instance._metadata_request = requests
+
+            return instance
+
+        # Now we set the relevant attributes of the function so that it seems
+        # like a normal method to the end user, with known expected arguments.
+        func.__name__ = f"set_{self.name}_request"
+        params = [
+            inspect.Parameter(
+                name="self",
+                kind=inspect.Parameter.POSITIONAL_OR_KEYWORD,
+                annotation=owner,
+            )
+        ]
+        params.extend(
+            [
+                inspect.Parameter(
+                    k,
+                    inspect.Parameter.KEYWORD_ONLY,
+                    default=UNCHANGED,
+                    annotation=Optional[Union[RequestType, str]],
+                )
+                for k in self.keys
+            ]
+        )
+        func.__signature__ = inspect.Signature(
+            params,
+            return_annotation=owner,
+        )
+        doc = REQUESTER_DOC.format(method=self.name)
+        for metadata in self.keys:
+            doc += REQUESTER_DOC_PARAM.format(metadata=metadata, method=self.name)
+        doc += REQUESTER_DOC_RETURN
+        func.__doc__ = doc
+        return func
+
+
+class _MetadataRequester:
+    """Mixin class for adding metadata request functionality.
+
+    .. versionadded:: 1.3
+    """
+
+    def __init_subclass__(cls, **kwargs):
+        """Set the ``set_{method}_request`` methods.
+
+        This uses PEP-487 [1]_ to set the ``set_{method}_request`` methods. It
```

Comment:
Third party libraries that inherit from `BaseEstimator` will get `set_{method}_requests` in their API docs. This will be confusing to the third party developers and their users.

I think we can use the feature flag to also control when `set_{method}_requests` is defined.

Reply:
This is run when the import statement is run. I don't think we want to attach/detach these methods when the config value is changed by the user. I'm not sure what it is that you're suggesting we do here.

Reply:
Ah yes, you are correct. Can we define the methods, but raise `AttributeError` when the feature flag is off?

Reply:
that doesn't really work, we're setting descriptors on classes. I thought we could do something like:

```py
            setattr(
                cls,
                f"set_{method}_request",
                _AvailableIfDescriptor(
                    RequestMethod(method, sorted(mmr.requests.keys())),
                    lambda obj: _routing_enabled(),
                    attribute_name=f"set_{method}_request"
                )
            )
```

which works for raising the `AttributeError`, but when you enable the flag, then calling the method gives:

```py
In [5]: SGDClassifier().set_fit_request()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[5], line 1
----> 1 SGDClassifier().set_fit_request()

File ~/Projects/sklearn/main/scikit-learn/sklearn/utils/_available_if.py:34, in _AvailableIfDescriptor.__get__(self, obj, owner)
     32     if not self.check(obj):
     33         raise attr_err
---> 34     out = MethodType(self.fn, obj)
     36 else:
     37     # This makes it possible to use the decorated method as an unbound method,
     38     # for instance when monkeypatching.
     39     @wraps(self.fn)
     40     def out(*args, **kwargs):

TypeError: first argument must be callable
```

So if we're to use our `avialable_if` descriptors, we'd need to add support for descriptors inside them. You know how? Otherwise we can simply add a check inside these methods and raise `AttributeError` if the flag is not enabled, or a `RuntimeError` probably would be better?

Reply:
Also, note that doing what you suggest, would mean in places where we do `set_fit_request` to be sure it's requested, no matter whether the flag is set or not, we'd need to add an if condition.

Reply:
Thinking it over, I think it's okay to not hide the method. For one to use metadata routing, they should have API docs to reference.

> would mean in places where we do set_fit_request to be sure it's requested,

Assuming we do not hide `set_fit_requests`, do you want `set_fit_request` to work when `enable_metadata_routing=False`? I think it makes sense to error. Imaging calling `set_fit_request` and passing it to a meta-estimator with the default `enable_metadata_routing=False`. The meta-estimator will use the old routing and ignore the `set_fit_request`.

> would mean in places where we do set_fit_request to be sure it's requested, no matter whether the flag is set or not, we'd need to add an if condition.

When I grep for `set_fit_request`, we only call it once in library code (`calibration`) on `LinearSVC`. In the normal cases, I see the user calling `set_fit_request`, which means they want routing and will set `enable_metadata_routing=True`.

Reply:
@thomasjpfan I can get behind the idea of `set_fit_request` raising if the flag is not enabled, but how do you propose we make this work?


```py
@pytest.mark.parametrize(
    "estimator, is_default_request",
    [
        # This fails, since the flag is not True by default. Pytest fails while collecting tests.
        (LinearRegression().set_fit_request(sample_weight=True), False),
        ...
    ],
)
def test_is_default_request(estimator, is_default_request):
    ...
```

Reply:
I think that is the only case where you call `set_fit_request` during test collection. I think you can work around this by:

```python
@pytest.mark.parametrize(
    "estimator, fit_request_kwargs, is_default_request",
    [
        (LinearRegression(), None, True),
        (LinearRegression(), dict(sample_weight=True), False),
        (WeightedMetaRegressor(estimator=LinearRegression()), None, True),
        (
            WeightedMetaRegressor(estimator=LinearRegression()),
            dict(sample_weight=True),
            False,
        ),
        (
            WeightedMetaRegressor(estimator=LinearRegression()),
            dict(sample_weight=True),
            False,
        ),
    ],
)
def test_is_default_request(estimator, fit_request_kwargs, is_default_request):
    if fit_request_kwargs:
        estimator.set_fit_request(**fit_request_kwargs)

    assert estimator.get_metadata_routing()._is_default_request == is_default_request
```

For the rest of the test have a `pytest.fixture` to auto configure them?

```python
@pytest.fixture(autouse=True)
def configure_slep006():
    orig_config = get_config()

    new_config = orig_config.copy()
    new_config["enable_metadata_routing"] = True
    set_config(**new_config)
    yield
    set_config(**orig_config)
```

Reply:
Done. Calling those methods now raises if feature flag is not enabled: https://github.com/scikit-learn/scikit-learn/pull/24027/commits/c3d4e4c80d390ab841ee8b66c888a246350f0419